ARROW-2822: [C++] Zero padding bytes in PoolBuffer

alendit opened the pull request:
See the [JIRA discussion](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-2822).

pitrou commented on cpp/src/arrow/builder.cc:
```diff
@@ -43,6 +44,33 @@ namespace arrow {
+
+ArrayBuilder::~ArrayBuilder() {
+#ifndef NDEBUG
+  if (ARROW_PREDICT_FALSE(!is_finished_)) {
```
Is this desirable? If an error occurs while building an array, it is normal not to finish the builder.

wesm replied:
How about hiding this behind a flag like `ARROW_LOG_UNFINISHED_BUILDERS`. Having this output in debug builds always seems like too much to me

alendit replied:
You want to rework it and move the warning to `null_bitmap` and `data` methods. It seems to be doable, but I'd need some more time. 

We can still then move it behind a special compile flag.

pitrou commented on cpp/src/arrow/builder.cc:
```diff
@@ -43,6 +44,33 @@ namespace arrow {
+    // zero the padding
+    buffer->ZeroPadding();
+  } else {
+    DCHECK_EQ(bytes_filled, 0);
```
Note null buffers are problematic, see PR #2243.

wesm replied:
I just left a review + comments there. I think we should try to merge this before merging #2243 so we can decide what to do there

pitrou commented on cpp/src/arrow/buffer.h:
```diff
@@ -101,6 +101,14 @@ class ARROW_EXPORT Buffer {
   Status Copy(const int64_t start, const int64_t nbytes,
               std::shared_ptr<Buffer>* out) const;
 
+  /// Zero bytes between padding, i.e. bytes between size_ and capacity_.
```
Not a native English speaker, but "between padding" looks grammatically incorrect.

wesm replied:
"in" instead of "between" sounds fine

alendit replied:
Yeah, comment editing fail, I'll fix it.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -222,7 +254,12 @@ void ArrayBuilder::UnsafeSetNotNull(int64_t length) {
+    RETURN_NOT_OK(TrimBuffer(BitUtil::BytesForBits(length_), null_bitmap_.get()));
+  }
+
+  *out = ArrayData::Make(null(), length_, {null_bitmap_}, length_);
```
What's the rationale for this change? All values in NullBuilder are null

alendit replied:
Oh, then I've misunderstood it's purpose. I thought it should be used as in `python_to_arrow.cc`. 

I'll revert it and add some unit cases, you could look at.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -1017,6 +1039,16 @@ DictionaryBuilder<T>::DictionaryBuilder(const std::shared_ptr<DataType>& type,
 
+template <typename T>
+DictionaryBuilder<T>::~DictionaryBuilder() {
+#ifndef NDEBUG
```
Ditto re: decision above

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -875,7 +897,7 @@ Status BooleanBuilder::AppendValues(const std::vector<bool>& values) {
                                  [values, &i]() -> bool { return values[i++]; });
 
-  // this updates length_
+  // this updates length_ and sets unfinished
```
Is setting unfinished here needed?

alendit replied:
Since unfinished signals that there were values appended after the last `FinishInternal` call, I would say yes.

wesm commented on cpp/src/arrow/builder.cc:
```diff
@@ -1149,15 +1181,20 @@ Status DictionaryBuilder<T>::Append(const Scalar& value) {
 
-Status DictionaryBuilder<NullType>::AppendNull() { return values_builder_.AppendNull(); }
+Status DictionaryBuilder<NullType>::AppendNull() {
+  is_finished_ = false;
```
Curious why this is needed

alendit replied:
There are nulls appended to the `values_builder_`, so it could resized and contain uninitialized padding, if we never call finish on it.

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -129,14 +130,15 @@ class ARROW_EXPORT ArrayBuilder {
       ++null_count_;
     }
     ++length_;
+    is_finished_ = false;
```
Seems like this should only be set when the builder is initialized

alendit replied:
This was my first try. Then I hit the problem that sometimes the builders are initialized but never appended to and never finished, i.e. in DictionaryBuilder. So I changed the logic to "set to false on every append". 

I'll see if I can change it back after the rework I've mentioned before.

wesm commented on cpp/src/arrow/io/file.cc:
```diff
@@ -199,6 +199,7 @@ class ReadableFile::ReadableFileImpl : public OSFile {
     RETURN_NOT_OK(Read(nbytes, &bytes_read, buffer->mutable_data()));
     if (bytes_read < nbytes) {
       RETURN_NOT_OK(buffer->Resize(bytes_read));
+      buffer->ZeroPadding();
```
It seems error prone to have to remember to zero the padding. What would be the performance consequences (if any, realistically) of zeroing in Resize?

pitrou replied:
Some things may be repeatedly resized if you don't know the exact size up front (e.g. in BinaryBuilder).

wesm replied:
Right, in the event that you shrank the buffer in `BinaryBuilder` when `Finish` is called, you would want to zero the padding there, too

wesm replied:
I guess if you shrink the memory then the memory you padded before goes unused

alendit replied:
The worst case is when you Reserve a huge amount of memory and then gradually Resize it, as the new elements come in. You would expect these operations to be virtually free while in reality you are zeroing everything between size and capacity on every call.

wesm replied:
Got it. OK, let's leave things as they are, then

wesm commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -279,10 +279,13 @@ class SequenceBuilder {
 
+    std::shared_ptr<Array> nones_array;
+    RETURN_NOT_OK(nones_.Finish(&nones_array));
+    const auto& nones = checked_cast<const NullArray&>(*nones_array);
```
the way that `NullBuilder` is being used in this module doesn't look right; probably related to the NullBuilder changes above cc @pcmoritz. It looks like a BooleanBuilder should be used instead

alendit replied:
I reverted the `NullBuilder` changes and switched `nones_` to `BooleanBuilder`.

@pcmoritz could you check if 6753dcc is fine?

alendit commented:
Hi guys. I'll try to incorporate the changes and rework the warning logic to be more precise in the comming days (hopefully tomorrow).

alendit commented:
Ok, so there was a possibility I totally haven't considered before: we can just remove the `data()` and `null_bitmap` from the builders. Amazingly, it doesn't break much besides couple of tests and the `DictionaryBuilder`. I adjusted `DictionaryBuilder` and it seems to work like a charm.

Does anyone know, if some 3rd party code relies on these methods? I would argue, they should be considered implementation detail, but we might want to put a deprecation warning before removing them alltogether.

pitrou commented:
> Does anyone know, if some 3rd party code relies on these methods? I would argue, they should be considered implementation detail, but we might want to put a deprecation warning before removing them alltogether.

I would deprecate them first indeed.

alendit commented:
Rebased on the master.

pitrou commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -545,24 +542,16 @@ class ORCFileReader::Impl {
-    uint8_t* target = reinterpret_cast<uint8_t*>(builder->data()->mutable_data());
+    std::vector<uint8_t> converted(source, source + length);
+
+    RETURN_NOT_OK(builder->AppendValues(converted.data(), length, valid_bytes));
```
Hmm... I think it's a bit problematic that we seem to be introducing a second data copy here (first from `source` to `converted`, then from `converted` to the builder). Can we avoid the second copy without using `builder->data()`? Otherwise we may have to undeprecate the latter.

wesm replied:
Note we don't have any unit tests for this code path so have to be extra careful

pitrou commented on cpp/src/arrow/buffer.h:
```diff
@@ -365,7 +374,7 @@ class ARROW_EXPORT TypedBufferBuilder : public BufferBuilder {
 };
 
-/// \brief Allocate a fixed-size mutable buffer from a memory pool
+/// \brief Allocate a fixed size mutable buffer from a memory pool, zero its padding.
```
I think "fixed-size" is the correct spelling. Perhaps a native English speaker can confirm?

wesm replied:
I googled a bit.  "fixed size" seems more common, and reads OK to me

pitrou commented on cpp/src/arrow/buffer.h:
```diff
@@ -384,7 +393,7 @@ Status AllocateBuffer(MemoryPool* pool, const int64_t size, std::shared_ptr<Buff
 Status AllocateBuffer(const int64_t size, std::shared_ptr<Buffer>* out);
 
-/// \brief Allocate a resizeable buffer from a memory pool
+/// \brief Allocate resizeable buffer from a memory pool, zero its padding.
```
I think the article here ("a resizeable buffer") shouldn't be removed.

wesm commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -580,17 +569,18 @@ class ORCFileReader::Impl {
     const int64_t* nanos = batch->nanoseconds.data() + offset;
-    int64_t* target = reinterpret_cast<int64_t*>(builder->data()->mutable_data());
 
+    std::vector<int64_t> values(length);
```
stack allocation here

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -68,7 +68,7 @@ class ARROW_EXPORT ArrayBuilder {
         capacity_(0) {}
 
-  virtual ~ArrayBuilder() = default;
+  virtual ~ArrayBuilder() {}
```
Are these lines equivalent (I thought so)?

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -566,9 +578,11 @@ class ARROW_EXPORT BooleanBuilder : public ArrayBuilder {
     return Status::OK();
   }
 
+  ARROW_DEPRECATED("Use Finish instread")
```
typo

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -357,9 +365,11 @@ class ARROW_EXPORT AdaptiveIntBuilderBase : public ArrayBuilder {
     return Status::OK();
   }
 
+  ARROW_DEPRECATED("Use Finish instread")
```
typo

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -214,8 +216,13 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
     return Status::OK();
   }
 
+  ARROW_DEPRECATED("Use Finish instread")
```
typo

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -1035,7 +1055,7 @@ class ARROW_EXPORT DictionaryBuilder : public ArrayBuilder {
 class ARROW_EXPORT DictionaryBuilder<NullType> : public ArrayBuilder {
  public:
-  ~DictionaryBuilder() override;
+  ~DictionaryBuilder() override {}
```
If this is trivial, it can be omitted altogether?

wesm commented on cpp/src/arrow/python/python_to_arrow.cc:
```diff
@@ -79,7 +79,7 @@ class SequenceBuilder {
     RETURN_NOT_OK(offsets_.Append(0));
     RETURN_NOT_OK(types_.Append(0));
-    return nones_.AppendToBitmap(false);
+    return nones_.AppendNull();
```
@pcmoritz or @robertnishihara take a look to make sure these changes look right. 

Should this be `Append(false)`? I would expect `nones_` to be an all non-null boolean array

wesm commented on python/pyarrow/tests/test_builder.py:
```diff
@@ -55,3 +55,5 @@ def test_string_builder_append_after_finish():
     expected = [None, None, "text", None, "other text"]
     assert arr.to_pylist() == expected
+    # prevent warning about unfinished builder
+    sbuilder.finish()
```
Is this still needed?

alendit commented:
So, what orc adapter did was to use `std::copy` to copy data from `int64_t*` input to a `Int32` builder by specifying `data()` as a target. This works because `std::copy` takes iterators and they just need to conform to the interator interface. IThe builders, on the other hand, either accept a vector or a raw pointer, both have to have the exact expected type to work.

My solution is to overload the `AppendValues` method to accept iterators. Then we can just use `std::copy` internally. The number of existing overloads with all possible combination of pointer and vector arguments suggests that we might want to switch to iterator-accepting interfaces at some point in the future, since they are more general. In the ideal world we'd have ranges, but alas...

An additional problem was that orc adapter used data pointer to directly add a computed value (time in nanoseconds). What is easily done in python with generators, is not supported in any current C++ version directly (excluding non-standard extensions). I added the `LazyRange`, a thin wrapper around a callable (e.g. lambda) which returns objects conforming to the iterator interface. What looks like a complicated construct, can be optimized by the compiler very well (I added some benchmarks). 

I really think we should remove/deprecate `data()` and `null_bitmap()` from the builders, but it's apparent, we'd need some interface restructuring. I think the changes would be beneficial in the long term, but I'm also not happy about the size of this diff. I would suggest that as long as the general direction is acceptable, we fix this PR's code and leave the more farreaching consideration to be fixed in further PRs.

alendit commented:
Ok, so funny thing: MSVC14 (i.e. 2015) issues warning on narrowing conversion when using `std::copy`. The funny part is that none other compiler, not even later MSVC versions, seem to be bothered by it. 

Which raises the question, if the orc adapter in its current form has even been through CI, because [here|https://github.com/apache/arrow/blob/master/cpp/src/arrow/adapters/orc/adapter.cc#L535] it totaly copies `int64_t` to `int32_t` and smaller. Of course, it's not a problem, since we check the supplied type beforehand, but it shouldn't compile with MSVC 2015.

With the supplied changes it's actually pretty easy to fix the warning: just wrap the iterator with a lambda which does a `static_cast` to the right type. Should I do it? Should be 0 costs when compiled on `Release`.

pitrou commented:
@alendit can you confirm when the PR is ready for review again? Thanks for working on this!

wesm commented:
>  I would suggest that as long as the general direction is acceptable, we fix this PR's code and leave the more farreaching consideration to be fixed in further PRs.

Sounds good to me

wesm commented:
Reviewing this again now

wesm commented on cpp/src/arrow/adapters/orc/adapter.cc:
```diff
@@ -521,18 +522,17 @@ class ORCFileReader::Impl {
-    target_type* target = reinterpret_cast<target_type*>(builder->data()->mutable_data());
+    auto cast_iter = internal::MakeLazyRange(
+        [&source](int64_t index) { return static_cast<target_type>(source[index]); },
+        length);
```
Cool

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -139,6 +144,38 @@ class ARROW_EXPORT ArrayBuilder {
+    int64_t bit_offset = length_ % 8;
+    uint8_t bitset = null_bitmap_data_[byte_offset];
+
+    for (auto iter = begin; iter != end; ++iter) {
```
As an aside, would you expect the optimized code to be significantly different here vs. the original `const uint8_t* valid_bytes` version?

alendit replied:
With STL iterators and raw pointers it should be exactly the same. With more complicated iterators, it's hard to tell, though the `LazyRange` ones the same performance as loops, at least on GCC.

pitrou commented on cpp/src/arrow/util/lazy-benchmark.cc:
```diff
@@ -0,0 +1,123 @@
+
+BENCHMARK(BM_lazy_copy)->Repetitions(3)->Unit(benchmark::kMillisecond);
+
+// std::copy with a lazy iterator which does static cast
```
I am curious: what do you call a lazy iterator? ISTM iterators are always lazy.

wesm replied:
I believe he's referring to the object returned by `internal::MakeLazyRange`

alendit replied:
Yes, the `LazyRange` is lazy about calling the supplied generator function instead of generating a new container eagerly like `std::generate`, for example. 

With these benchmarks I wanted to make sure, that the compilers can inline the lambdas properly so we don't sacrifice performance compared to a loop.

wesm commented on cpp/src/arrow/builder.h:
```diff
@@ -214,8 +249,13 @@ class ARROW_EXPORT PrimitiveBuilder : public ArrayBuilder {
+  ARROW_DEPRECATED("Use Finish instead")
   std::shared_ptr<Buffer> data() const { return data_; }
 
+  const value_type GetValue(int64_t index) const {
```
Since these are always C types I would think that just `value_type` is fine; I am not an expert on the fine details, though

alendit replied:
Yupp, the first const should go.

wesm commented:
:+1: ok, I'm merging so we can get your other PR rebased. Thanks @alendit @pitrou for your careful attention on this issue

alendit commented:
Nice, glad it went through. Thx both of you for the help and patience :) 

Once further question: for really small changes, like removing that one `const`, do I open a JIRA issue, too? Or is there a simplified procedure, so to not to swamp the queue?

wesm commented:
You can open a PR but don't bother creating a JIRA. I'll merge once the build runs
